File: IEEE Practice/waterloo problems/coin_collector.cpp


#include <iostream>

bool * calc_change(unsigned int value, unsigned int k, unsigned int* coins, unsigned int n);

int main()
{
	unsigned int k, n;
	std::cin >> n >> k;
	unsigned int *coins = new unsigned int(n);
	bool* collection = new bool(n);

	for (unsigned int i = 0; i < n; i++) {
		std::cin >> coins[i];
		int tmp;
		std::cin >> tmp;
		if (tmp > 0)
			collection[i] = true;
		else
			collection[i] = false;
	}
	int max_new_coins = 0;
	unsigned int max_value = 0;

	for (int i = k - 1; i > 0; i--) {
		int new_coins = 0;
		bool* change = calc_change(i, k, coins, n);
		for (unsigned int j = 0; j < n; j++) {
			if (change[j] && !collection[j])
				new_coins++;
		}
		if (new_coins > max_new_coins) {
			max_new_coins = new_coins;
			max_value = i;
		}
	}
	std::cout << max_new_coins << std::endl << max_value;
	return(0);
}

bool* calc_change(unsigned int value, unsigned int k, unsigned int* coins, unsigned int n) {
	bool* ans = new bool(n);
	unsigned int change = k - value;
	for (int i = n - 1; i >= 0; i--) {
		if (coins[i] <= change) {
			ans[i] = true;
			change = change % coins[i];
		}
		else
			ans[i] = false;
	}
	return ans;
}

File: IEEE Practice/waterloo problems/coin_collector.py
import sys

input = sys.stdin.read().split('\n')

data = []

for i in range(len(input)):
    data.append(input[i].split(' ')

print (data)


'''
n = int(data[0][0])
K = int(data[0][1])

coins =[]
collection = []

for i in range(1,len(data)):
    coins.append(int(data[i][0]))
    if int(data[i][1]) > 0:
        collection.append(True)
    else:
        collection.append(False)


K=25
coins = [1,2,3,5,10,13,20]
collection = [False, False, True, False, False,False,False]

def calc_change (value, K, coins):
    ans = []
    change = K - value
    for i in range (len(coins)-1, -1, -1):
        if coins[i] <= change:
            ans.append(coins[i])
            change = change%coins[i]
    return ans

def find_num_new_coins (change_array, collection):
    num_new =0
    for i in range (len(coins)):      
        if (coins[i] in change_array):
            if (not collection[i]):
                num_new = num_new +1
    return num_new

max_new_coins = 0
max_value = 0

for i in range (K-1, 0, -1):
    change = calc_change(i, K, coins)
    new_coins = find_num_new_coins(change, collection)

    if new_coins > max_new_coins:
        max_new_coins = new_coins
        max_value = i

print (max_new_coins, "\n", max_value)
'''

